### PR TITLE
Complete basic testing for IRISpy

### PR DESCRIPTION
**Pull request by kakirastern**
Aim 
===
To complete the basic tests for IRISpy to run smoothly.

### CODE REVIEW

**Comment by kakirastern** [comment]
Continuing work on #117.

**Comment by kakirastern** [comment]
Working on getting the existing tests passed now...

Screenshot 2019-08-08 23 38 37

File: irispy/tests/test_iris_tools.py [hunk 1]
```diff
@@ -157,7 +158,8 @@ def test_uncalculate_exposure_time_correction_error(input_arrays, old_unit, expo
                                                                exposure_time, force=force)
 
 def test_get_iris_response_not_equal_to_one():
-    assert pytest.raises(TypeError, iris_tools.get_iris_response, time_obs, pre_launch=False, response_version=3)
+    assert pytest.raises(KeyError, iris_tools.get_iris_response, time_obs,
+                         pre_launch=False, response_version=13)
```
Comment (DanRyanIrish) [inline irispy/tests/test_iris_tools.py]: Not sure this test is performing the same task now that it did before...

**Comment by kakirastern** [comment]
Screenshot 2019-08-09 00 28 00

One final hurdle to surpass before all existing local tests are passed: 

```
_____ test_get_iris_response_version4[input_quantity0-expected_quantity0] ______

input_quantity = array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])
expected_quantity = array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]], dtype=float32)

    @pytest.mark.parametrize("input_quantity, expected_quantity",
    [
     (iris_response4["AREA_SG"].value, area_sg_load4),
     (iris_response4["AREA_SJI"].value, area_sji_load4)
     ])
    
    def test_get_iris_response_version4(input_quantity, expected_quantity):
>       np_test.assert_almost_equal(input_quantity, expected_quantity, decimal=4)
E       AssertionError: 
E       Arrays are not almost equal to 4 decimals
E       
E       Mismatch: 1.49%
E       Max absolute difference: 0.25385037
E       Max relative difference: nan
E        x: array([[0., 0., 0., ..., 0., 0., 0.],
E              [0., 0., 0., ..., 0., 0., 0.]])
E        y: array([[0., 0., 0., ..., 0., 0., 0.],
E              [0., 0., 0., ..., 0., 0., 0.]], dtype=float32)

irispy/tests/test_iris_tools.py:285: AssertionError
```
This error is strange, because it might have to do with the `time_obs` date chosen which could have introduced some errors in `iris_response["AREA_SG"]`... But I am not sure. 

Could you confirm you see this error as well @DanRyanIrish?

File: irispy/tests/test_iris_tools.py (hunk 1 above)
Comment (kakirastern) [reply irispy/tests/test_iris_tools.py]: No, I meant to ask you if you want it removed. It was there before version 3 was incorporated so I used a creative way to sort of "commenting it out." Would you like to keep it with some modifications?

**Comment by kakirastern** [comment]
And some additional sleuth work yields the following: 

```
In [13]: sum(np.array(area_sg_load4[0]))                                        
Out[13]: 151.58501255512238

In [14]: sum(np.array(area_sg_load4[1]))                                        
Out[14]: 26.30692084133625

In [15]: np.sum(iris_response['AREA_SG'].value[0])                              
Out[15]: 151.5850126312236

In [16]: np.sum(iris_response['AREA_SG'].value[1])                              
Out[16]: 0.0

```

**Comment by kakirastern** [comment]
Or it might have been some confusion regarding the `.sav` file for version 4 we have been using...

**Comment by kakirastern** [comment]
Hi @DanRyanIrish I will try and find Laura and see if I could double check that `AREA_SG` data in the `.sav` file provided by her with her... Seems like something is off, but I have no clue what is wrong at this stage.

File: irispy/tests/test_iris_tools.py (hunk 1 above)
Comment (kakirastern) [reply irispy/tests/test_iris_tools.py]: It was for ruling out version `3` or `4` when they were not supposed to have been implemented in `get_iris_response`, so the test would throw a `KeyError` originally, I changed it to `TypeError` in a previous PR to pass the local tests, if my memory is any reliable. So this test may not be necessary any more.

**Comment by kakirastern** [comment]
Remaining bug fixed in `get_iris_response` for version 4. Now the matching between input and expected values for this version is much better: 

```
_____ test_get_iris_response_version4[input_quantity0-expected_quantity0] ______

input_quantity = array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])
expected_quantity = array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]], dtype=float32)

    @pytest.mark.parametrize("input_quantity, expected_quantity",
    [
     (iris_response4["AREA_SG"].value, area_sg_load4),
     (iris_response4["AREA_SJI"].value, area_sji_load4)
     ])
    
    def test_get_iris_response_version4(input_quantity, expected_quantity):
>       np_test.assert_almost_equal(input_quantity, expected_quantity, decimal=6)
E       AssertionError: 
E       Arrays are not almost equal to 6 decimals
E       
E       Mismatch: 1.39%
E       Max absolute difference: 0.00853752
E       Max relative difference: nan
E        x: array([[0., 0., 0., ..., 0., 0., 0.],
E              [0., 0., 0., ..., 0., 0., 0.]])
E        y: array([[0., 0., 0., ..., 0., 0., 0.],
E              [0., 0., 0., ..., 0., 0., 0.]], dtype=float32)

irispy/tests/test_iris_tools.py:285: AssertionError
```

**Comment by kakirastern** [comment]
But the agreement is good up to 2 decimal places for version 4... Must be the way I wrote the coded for the spline fit: 

Screenshot 2019-08-10 03 07 42

**Comment by kakirastern** [comment]
I have tested different ways to carry out the `spline` fit, and have settled with the following implementation: 

```python
tck = scipy.interpolate.splrep(iris_response["C_N_LAMBDA"].value, np.squeeze(iris_fit_nuv[k, :]), s=0, k=3)
iris_response["AREA_SG"][1, w_nuv] = scipy.interpolate.splev(iris_response["LAMBDA"][w_nuv].value, tck, ext=0)
```

**Comment by kakirastern** [comment]
Hi @DanRyanIrish! Any suggestions on how I could improve the accuracy of the `spline` fit with that produced with the `IDL`? I can learn a lot from your expertise in this area.

File: irispy/iris_tools.py [hunk 2]
```diff
@@ -257,15 +261,19 @@ def get_iris_response(time_obs=None, pre_launch=False, response_file=None, respo
         for j in range(shp_nuv[0]):
             iris_fit_nuv[:, j] = fit_iris_xput(time_obs, iris_response["C_N_TIME"], iris_response["COEFFS_NUV"][j, :, :])
             # Interpolate onto lambda grid
-            w_nuv = np.where(np.logical_and(iris_response["LAMBDA"]/u.nm >= lambran_nuv[0], iris_response["LAMBDA"]/u.nm <= lambran_nuv[1]))
+            w_nuv = np.where(np.logical_and(iris_response["LAMBDA"].value >= lambran_nuv[0], iris_response["LAMBDA"].value <= lambran_nuv[1]))
```
Comment (hayesla) [inline irispy/iris_tools.py]: this shouldnt be indented

File: irispy/iris_tools.py [hunk 3]
```diff
@@ -257,15 +261,19 @@ def get_iris_response(time_obs=None, pre_launch=False, response_file=None, respo
...
         if int(iris_response["VERSION"]) <= 3:
             for k in range(n_time_obs):
                 interpol_nuv =  scipy.interpolate.interp1d(iris_response["C_N_LAMBDA"][:], np.squeeze(iris_fit_nuv[k, :]), fill_value='extrapolate')
                 iris_response["AREA_SG"][1, w_nuv] = interpol_nuv(iris_response["LAMBDA"][w_nuv])
         else:
             for k in range(n_time_obs):
```
Comment (hayesla) [inline irispy/iris_tools.py]: else:
            for k in range(n_time_obs):
                interpol_nuv = scipy.interpolate.CubicSpline(iris_response["C_N_LAMBDA"][:], np.squeeze(iris_fit_nuv[k, :]), extrapolate=True)
                iris_response['AREA_SG'][1, w_nuv] = interpol_nuv(iris_response["LAMBDA"][w_nuv])

File: irispy/iris_tools.py [hunk 4]
```diff
@@ -256,16 +260,22 @@ def get_iris_response(time_obs=None, pre_launch=False, response_file=None, respo
         detector_nuv = "NUV"
         for j in range(shp_nuv[0]):
             iris_fit_nuv[:, j] = fit_iris_xput(time_obs, iris_response["C_N_TIME"], iris_response["COEFFS_NUV"][j, :, :])
-            # Interpolate onto lambda grid
-            w_nuv = np.where(np.logical_and(iris_response["LAMBDA"]/u.nm >= lambran_nuv[0], iris_response["LAMBDA"]/u.nm <= lambran_nuv[1]))
+                # Interpolate onto lambda grid
+                w_nuv = np.where(np.logical_and(iris_response["LAMBDA"].value >= lambran_nuv[0], iris_response["LAMBDA"].value <= lambran_nuv[1]))
```
Comment (hayesla) [inline irispy/iris_tools.py]: this is the wrong way - you want it out of the loop

**Comment by kakirastern** [comment]
Hi @DanRyanIrish So after consulting with @hayesla and a bit of experimentation, am able to drive the max abs error down to `0.0833%` or `3 decimal places`:

```
_____ test_get_iris_response_version4[input_quantity0-expected_quantity0] ______

input_quantity = array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])
expected_quantity = array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]], dtype=float32)

    @pytest.mark.parametrize("input_quantity, expected_quantity",
    [
     (iris_response4["AREA_SG"].value, area_sg_load4),
     (iris_response4["AREA_SJI"].value, area_sji_load4)
     ])
    
    def test_get_iris_response_version4(input_quantity, expected_quantity):
>       np_test.assert_almost_equal(input_quantity, expected_quantity, decimal=4)
E       AssertionError: 
E       Arrays are not almost equal to 4 decimals
E       
E       Mismatch: 0.0833%
E       Max absolute difference: 0.00020058
E       Max relative difference: nan
E        x: array([[0., 0., 0., ..., 0., 0., 0.],
E              [0., 0., 0., ..., 0., 0., 0.]])
E        y: array([[0., 0., 0., ..., 0., 0., 0.],
E              [0., 0., 0., ..., 0., 0., 0.]], dtype=float32)

irispy/tests/test_iris_tools.py:285: AssertionError
===================== 1 failed, 108 passed in 4.99 seconds =====================
```

Is this margin of error tolerable or acceptable for the purposes of the cubic spline fit? Please advise.

File: irispy/iris_tools.py [hunk 5]
```diff
@@ -347,8 +357,8 @@ def fit_iris_xput(time_obs, time_cal_coeffs, cal_coeffs):
 
     Parameters
     ----------
-    time_obs: a numpy array of floats
-    - Times of datapoints.
+    time_obs: a `numpy.array`
+        A set of observation times contained in a numpy array.
```
Comment (DanRyanIrish) [inline irispy/iris_tools.py]: Say that each element in the array must be an astropy Time object.

Comment (kakirastern) [reply irispy/iris_tools.py]: But for the `fit_iris_xput` the `time_obs` is indeed a numpy array.

Comment (hayesla) [reply irispy/iris_tools.py]: but you need to say that each element in the array have to be an astropy Time object - i.e. say that the set of observation times are astropy times

Comment (kakirastern) [reply irispy/iris_tools.py]: Okay, will do

**Comment by DanRyanIrish** [comment]
Merging this now.  Well done @kakirastern!

**Pull request merged**